svelte/src/scripts/draw/helper/tree-layout.ts

```diff
@@ -0,0 +1,142 @@
+	treeData?: {
+		targets: Map<simluationNodeDatumType, number>; // Store nodes at lower level
+		level: number; // The depth of the node in the tree
+	}
+};
+
+/** 
+ * Caculates a set of arboricities on the given graph, but only on the top level in the dependency hierarchy.
+ * (Not necicarilly maximal or efficient)
+ * 
+ * Stores data in a new property treeData on the node (in place)
+ * 
+ * TODO move to parser
+ * */ 
+function findArboricity(nodes: simluationNodeDatumTypeExt[]) {
+	// Helper functions
+	function getOutLinks(node: simluationNodeDatumTypeExt): simluationLinkType[] {
+		return node.members.flatMap(m => getOutLinks(m))
+		.concat(node.outgoingLinks ?? []);
+	};
+
+	function getCompoundNode(node: simluationNodeDatumType): simluationNodeDatumType {
+		if (nodes.includes(node)) {
+			return node;
+		} else if (node.parent) {
+			return getCompoundNode(node.parent);
+		} else {
+			throw(`Invalid node data, ${node.id} is not related to any node passed to findArboricity`);
+		}
+	}
+
+	// (Works in place!)
+	function liftOutLinks(node: simluationNodeDatumTypeExt, links: simluationLinkType[]) {
+		links.forEach((link) => {
+			const outNode = getCompoundNode(link.target);
+			if (outNode !== node) {
+				node.treeData!.targets!.set(outNode, (node.treeData!.targets!.get(outNode) ?? 0) + 1);
+			}
+		});
+	}
+
+	// Start actual logic
+	// Prepare data/structure
+	// All dependencies must be lifted, but not in render data
+	nodes.forEach((node) => {
+		node.treeData = {
+			targets: new Map(),
+			level: NaN,
+		};
+		const links = getOutLinks(node);
+		liftOutLinks(node, links);
+	});
+
+	// Get node with maximum outgoing connections, we assume this is the root of the first arboricity.
+	const findRootReduce = (a: simluationNodeDatumTypeExt, node: simluationNodeDatumTypeExt) => 
+		node.treeData!.targets.size > a.treeData!.targets.size ? node : a;
+	const initialRoot = nodes.reduce(findRootReduce);
+	initialRoot.treeData!.level = 0;
+
+	// Set up hull data
+	const nodesCovered = new Set<simluationNodeDatumTypeExt>();
+	const levelMap = new Map<number, Set<simluationNodeDatumTypeExt>>(); // Necacary for calculating force strength eventually.
+
+	/**
+	 * Helper function, to caculate arboricity/tree structure (Greedy algorithm)
+	 * Assumes everything is connected
+	 * */
+	function calculateTree(node: simluationNodeDatumTypeExt, level: number): void {
+		// Add current node to cover
+		node.treeData!.level = level;
+		if (levelMap.has(level)) {
+			levelMap.get(level)!.add(node) 
+		} else {
+			levelMap.set(level, new Set());
+		};
+		nodesCovered.add(node);
+
+		let options : {weigth: number; target: simluationNodeDatumTypeExt; level: number}[] = [];
+		do {
+			options = [];
+			nodesCovered.forEach((n) => {
+				n.treeData!.targets
+				.forEach((weigth, target) => {
+					if (!nodesCovered.has(target)) {
+						const o = options.find((o) => o.target === target);
```

Comment:
dang the indentation is quite deep here lol

Reply:
might want to refactor it?

Reply:
True, but the lines aren't wide and we're probably not reusing tree stuff anywhere.